pkg/volume/aws_ebs/aws_ebs.go

```
@@ -241,6 +243,166 @@ func (plugin *awsElasticBlockStorePlugin) ConstructVolumeSpec(volName, mountPath
+}
+
+func (plugin *awsElasticBlockStorePlugin) newBlockVolumeMapperInternal(spec *volume.Spec, podUID types.UID, manager ebsManager, mounter mount.Interface) (volume.BlockVolumeMapper, error) {
+	ebs, readOnly, err := getVolumeSource(spec)
+	if err != nil {
+		return nil, err
+	}
+	volumeID := aws.KubernetesVolumeID(ebs.VolumeID)
+	fsType := ebs.FSType
+	partition := ""
+	if ebs.Partition != 0 {
+		partition = strconv.Itoa(int(ebs.Partition))
+	}
+
+	return &awsElasticBlockStoreMapper{
+		awsElasticBlockStore: &awsElasticBlockStore{
+			podUID:          podUID,
+			volName:         spec.Name(),
+			volumeID:        volumeID,
+			partition:       partition,
+			manager:         manager,
+			mounter:         mounter,
+			plugin:          plugin,
+			MetricsProvider: volume.NewMetricsStatFS(getPath(podUID, spec.Name(), plugin.host)),
+		},
+		fsType:      fsType,
+		readOnly:    readOnly,
+		diskMounter: volumehelper.NewSafeFormatAndMountFromHost(plugin.GetPluginName(), plugin.host)}, nil
+}
+
+func (plugin *awsElasticBlockStorePlugin) NewBlockVolumeUnmapper(volName string, podUID types.UID) (volume.BlockVolumeUnmapper, error) {
+	return plugin.newUnmapperInternal(volName, podUID, &AWSDiskUtil{})
+}
+
+func (plugin *awsElasticBlockStorePlugin) newUnmapperInternal(volName string, podUID types.UID, manager ebsManager) (volume.BlockVolumeUnmapper, error) {
+	pluginDir := plugin.host.GetVolumeDevicePluginDir(awsElasticBlockStorePluginName)
+	vID := findVolumeIDFromPath(string(podUID), pluginDir)
+	volID := aws.KubernetesVolumeID(vID)
+
+	return &awsElasticBlockStoreUnmapper{
+		awsElasticBlockStore: &awsElasticBlockStore{
+			podUID:   podUID,
+			volumeID: volID,
+			volName:  volName,
+			manager:  manager,
+			plugin:   plugin,
+		}}, nil
+}
+
+func findVolumeIDFromPath(podUID string, pluginDir string) string {
+	files, err := ioutil.ReadDir(pluginDir)
+	if err != nil {
+		return ""
```

This method should return error if something is wrong instead of returning empty string, otherwise you'll hit another error when you use volumeID like removing globalMapPath.